src/main/scala/diplomacy/Nodes.scala

```diff
@@ -10,6 +10,90 @@ import freechips.rocketchip.util.HeterogeneousBag
 import scala.collection.mutable.ListBuffer
 import scala.util.matching._
 
+/**
+ * Concepts, metaphors, and mnemonics to help with understanding Diplomacy code.
+ *
+ *
+ * # Abstract types
+ *
+ * Diplomacy is a set of abstractions for describing directed, acyclic graphs
+ * where parameters may be negotiated between nodes. These abstractions are
+ * expressed in the form of abstract classes, traits, and type parameters, which
+ * comprise nearly all of the types defined in this file.
+ *
+ * The NodeImp ("node implementation") is the main abstract type that determines
+ * the type parameters of all other abstract types. Defining a concrete
+ * implementation of NodeImp will therefore determine concrete types for all
+ * type parameters. For example, passing in a concrete instance of NodeImp to a
+ * SourceNode will fully determine concrete types for all of a SourceNode's type
+ * parameters.
+ *
+ * Specific applications of Diplomacy are expected to either extend these types
+ * or to specify concrete types for the type parameters. This allows for
+ * creating application-specific node, edge, parameter, and bundle types.
+ *
+ * Is an "edge" and a binding the same thing?
+ *
+ *
+ * # Inward/Outward vs. Upward/Downward
+ *
+ * Diplomacy defines two dimensions: inward/outward and upward/downward.
+ *
+ * Inward/outward refer to the direction of the directed graph itself. For a
+ * given node:
+ * - Inward refers to edges that point into a node.
+ * - Outward refers to edges that point out of a node.
+ *
+ * A useful mnemonic for distinguishing between inward and outward is to always
+ * view it from the perspective of a particular node. Because Diplomacy
+ * describes directed, acyclic graphs, this direction will always be consistent
+ * across the entire graph.
+ *
+ * Upward/downward refer to the direction of the parameter negotiation, where
+ * the direction is relative to the inward/outward direction. For a given edge:
+ * - Upward refers to a flow of parameters in the outward direction.
+ * - Downward refers to a flow of parameters in the inward direction.
+ *
+ * A useful mnemonic for distinguishing between inward and outward is to imagine
+ * the diplomatic graph as a literal network of rivers where upward refers to
+ * parameters that move in the upstream direction while downward refers to
+ * parameters that move in the downstream direction.
+ *
+ *
+ * # Handles
+ *
+ * Two Diplomatic nodes can be bound together using the := operator or one of
+ * its sibling operators. Binding is asymmetric, and the binding operation will
```

Comment:
Why did you decide to put so much here about the `:=` operator, but not go into `:=*` and friends? Did you just run out of time or was that a concious division? Later we talk about "ostars" without any intro

Reply:
I was sort of thinking of mentioning the other operators, but I think that probably goes into a different section that describes the operators in more detail. My primary goal for writing this section was to understand what a `Handle` is, why it exists, and something to help with why it's specifically named a "handle". I'd also want to run everything I've written here by @terpstra, since everything I wrote is just me trying to "create a theory of handles" from reading the code, and I'm not sure if the design principles I post hoc derived are actually what @terpstra's goals were.

This Handle stuff is only really a concept that you'd need to understand if you were trying to deeply understand the code here, but it's not necessarily something you'd need to know as a user of a library implemented in Diplomacy. For example, anyone using the Diplomatic TileLink framework needs to understand `:=`, `:=*`, etc., but they wouldn't need to know anything about a Handle.

For what it's worth, the last part of the code that I got to on the plane while trying to was the four different `NodeBinding` types and their associated operators, but I had not gotten to the point of actually comprehending the code yet, since the code is recursive, meaning that in order to understand `resolveStar()`, you need to understand how `resolveStar()` works on the neighboring nodes. This is a bit challenging when you don't have internet or any prior knowledge on using any of the operators besides the plain "bind-once" `:=` operator.

Reply:
> My primary goal for writing this section was to understand what a Handle is, why it exists, and something to help with why it's specifically named a "handle".

I think handle is implemented as its name.
```
nodeA := nodeB
```
will return a new handle with `nodeA` as `outwardNode` and `nodeB` as `inwardNode`.
If `nodeB` doesn't have a `inwardNode`, `bind` will just return a `OutwardNodeHandle` which only have a `outwardNode` `nodeA`.